### PR TITLE
Add Triangle.getUV & The result of sprite.raycast add uv

### PR DESCRIPTION
**Pull request by 06wj**
Fixed https://github.com/mrdoob/three.js/pull/14292#issuecomment-415964766

### CODE REVIEW

File: src/objects/Sprite.js [hunk 1]
```diff
@@ -93,6 +98,22 @@ Sprite.prototype = Object.assign( Object.create( Object3D.prototype ), {
 
 		}
 
+		var barycoord = new Vector3();
+
+		function uvIntersection( point, p1, p2, p3, uv1, uv2, uv3 ) {
```
Comment (WestLangley) [inline src/objects/Sprite.js]:
```javascript
uvIntersection( point, p1, p2, p3, uv1, uv2, uv3, result )
```

File: src/objects/Sprite.js [hunk 2]
```diff
@@ -93,6 +98,22 @@ Sprite.prototype = Object.assign( Object.create( Object3D.prototype ), {
...
+
+			Triangle.getBarycoord( point, p1, p2, p3, barycoord );
+
+			uv1.multiplyScalar( barycoord.x );
```
Comment (WestLangley) [inline src/objects/Sprite.js]: Don't change input data.

File: src/objects/Sprite.js [hunk 3]
```diff
@@ -93,6 +98,22 @@ Sprite.prototype = Object.assign( Object.create( Object3D.prototype ), {
...
+			uv2.multiplyScalar( barycoord.y );
+			uv3.multiplyScalar( barycoord.z );
+
+			uv1.add( uv2 ).add( uv3 );
+
+			return uv1.clone();
```
Comment (WestLangley) [inline src/objects/Sprite.js]:
```javascript
return result;
```

File: src/objects/Sprite.js [hunk 4]
```diff
@@ -138,6 +165,7 @@ Sprite.prototype = Object.assign( Object.create( Object3D.prototype ), {
 
 				distance: distance,
 				point: intersectPoint.clone(),
+				uv: uvIntersection( intersectPoint, vA, vB, vC, uvA, uvB, uvC ),
```
Comment (WestLangley) [inline src/objects/Sprite.js]:
```javascript
uvIntersection( intersectPoint, vA, vB, vC, uvA, uvB, uvC ).clone();
```

File: src/objects/Sprite.js (hunk 3 above)
Comment (06wj) [reply src/objects/Sprite.js]: https://github.com/mrdoob/three.js/blob/dev/src/objects/Mesh.js#L143-L160

Is the code here should also be modified?

File: src/objects/Sprite.js (hunk 4 above)
Comment (makc) [reply src/objects/Sprite.js]: this will make sense if uvIntersection moved some place where it's public, and then usedin both sprite and mesh

Comment (WestLangley) [reply src/objects/Sprite.js]: I think it is OK if it is fixed in both places.

**Comment by mrdoob** [comment]
@WestLangley looks good?

File: src/math/Triangle.js [hunk 5]
```diff
@@ -106,6 +107,29 @@ Object.assign( Triangle, {
...
+			v1.copy( uv1 ).multiplyScalar( barycoord.x );
+			v2.copy( uv2 ).multiplyScalar( barycoord.y );
+			v3.copy( uv3 ).multiplyScalar( barycoord.z );
+
+			v1.add( v2 ).add( v3 );
+
+			return result.copy( v1 );
+
+		};
+
```
Comment (WestLangley) [inline src/math/Triangle.js]: How about naming the method `getUV()`. You can simplify the code like so:
```javascript
getUV: function () {

	var barycoord = new Vector3();

	return function getUV( point, p1, p2, p3, uv1, uv2, uv3, target ) {

		this.getBarycoord( point, p1, p2, p3, barycoord );

		target.set( 0, 0 );
		target.addScaledVector( uv1, barycoord.x );
		target.addScaledVector( uv2, barycoord.y );
		target.addScaledVector( uv3, barycoord.z );

		return target;

	};

}()
```

Comment (WestLangley) [reply src/math/Triangle.js]: @06wj I see you used this code.  Did you test it?

Comment (06wj) [reply src/math/Triangle.js]: Yes, I test it in [webgl_raycast_sprite](https://rawgit.com/06wj/three.js/patch1_build/examples/webgl_raycast_sprite.html) and [webgl_raycast_texture](https://rawgit.com/06wj/three.js/patch1_build/examples/webgl_raycast_texture.html)

Comment (WestLangley) [reply src/math/Triangle.js]: Great!

**Comment by WestLangley** [comment]
@mrdoob I think this is OK to merge.

@Mugen87 Can you think of further improvements?

**Comment by mrdoob** [comment]
Thanks!

**Comment by Mugen87** [comment]
I think the code is okay like.

**Pull request merged**